File: swift/Sources/Armenus/ArmenusModelView.swift
#if os(iOS)
import SceneKit
import UIKit

/// A dish in 3D, with AR where the handset supports it. UIKit.
///
/// SceneKit, not RealityKit. RealityKit spins up an ARView with a render loop
/// and a session per instance, which is heavy for a card in a scrolling list
/// that is not doing AR. `SCNView` is a plain view that draws a scene and stops
/// when it is off screen.
///
/// USDZ only. SceneKit has no glTF importer, which is why `EmbedModel.usdzUrl`
/// gates the preview on iOS and not only the AR button. The view never
/// second-guesses ``resolvePresentation(platform:model:arCoreAvailable:hasWebXR:isMobileWeb:)``:
/// what it draws is whatever that returned.
public final class ArmenusModelView: UIView {

  // MARK: Configuration

  public var arLabel: String = "View on your table" {
    didSet { applyButtonStyle() }
  }

  /// Whether a drag rotates the model. Pass false inside a scrolling list: on
  /// a phone the two gesture systems fight, and a card that swallows vertical
  /// drags makes the whole feed feel stuck.
  public var interactionEnabled: Bool = true {
    didSet { sceneView.allowsCameraControl = interactionEnabled }
  }

  /// Background of the AR button.
  public var accentColor: UIColor = UIColor(red: 0.137, green: 0.302, blue: 0.235, alpha: 1) {
    didSet { applyButtonStyle() }
  }

  public var onModelLoad: (() -> Void)?
  public var onModelError: ((String) -> Void)?
  public var onEnterAR: ((EmbedItem) -> Void)?

  public private(set) var item: EmbedItem?
  public private(set) var presentation = resolvePresentation(platform: .iOS, model: nil)

  // MARK: Subviews

  private let stage = UIView()
  private let sceneView = SCNView()
  private let posterView = UIImageView()
  private let spinner = UIActivityIndicatorView(style: .medium)
  private let arButton = UIButton(type: .system)
  private let noteLabel = UILabel()

  private var modelNode: SCNNode?
  private var currentSource: String?
  private var posterTask: Task<Void, Never>?
  private var presenting = false

  // MARK: Lifecycle

  public override init(frame: CGRect) {
    super.init(frame: frame)
    setUp()
  }

  public required init?(coder: NSCoder) {
    super.init(coder: coder)
    setUp()
  }

  private func setUp() {
    stage.backgroundColor = UIColor(red: 0.957, green: 0.949, blue: 0.953, alpha: 1)
    stage.layer.cornerRadius = 14
    stage.clipsToBounds = true

    posterView.contentMode = .scaleAspectFit

    sceneView.backgroundColor = .clear
    sceneView.antialiasingMode = .multisampling4X
    // Physically based lighting, so a dish lit for PBR does not come out flat.
    sceneView.autoenablesDefaultLighting = true
    sceneView.allowsCameraControl = interactionEnabled
    // Paused until a model is loaded, and paused when the view leaves the
    // window. Each SCNView owns a CADisplayLink; twenty dish cards rendering
    // off screen drain a battery and drop the scroll below 60fps for nothing.
    sceneView.rendersContinuously = false
    sceneView.isPlaying = false
    sceneView.isHidden = true

    spinner.hidesWhenStopped = true

    applyButtonStyle()
    arButton.isHidden = true
    arButton.addTarget(self, action: #selector(enterAR), for: .touchUpInside)

    noteLabel.font = .systemFont(ofSize: 13)
    noteLabel.textColor = .secondaryLabel
    noteLabel.textAlignment = .center
    noteLabel.numberOfLines = 0
    noteLabel.isHidden = true

    for view in [stage, noteLabel] {
      view.translatesAutoresizingMaskIntoConstraints = false
      addSubview(view)
    }
    for view in [posterView, sceneView, spinner, arButton] {
      view.translatesAutoresizingMaskIntoConstraints = false
      stage.addSubview(view)
    }

    NSLayoutConstraint.activate([
      stage.topAnchor.constraint(equalTo: topAnchor),
      stage.leadingAnchor.constraint(equalTo: leadingAnchor),
      stage.trailingAnchor.constraint(equalTo: trailingAnchor),
      stage.heightAnchor.constraint(equalTo: stage.widthAnchor),

      noteLabel.topAnchor.constraint(equalTo: stage.bottomAnchor, constant: 10),
      noteLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
      noteLabel.trailingAnchor.constraint(equalTo: trailingAnchor),
      noteLabel.bottomAnchor.constraint(equalTo: bottomAnchor),

      posterView.topAnchor.constraint(equalTo: stage.topAnchor),
      posterView.leadingAnchor.constraint(equalTo: stage.leadingAnchor),
      posterView.trailingAnchor.constraint(equalTo: stage.trailingAnchor),
      posterView.bottomAnchor.constraint(equalTo: stage.bottomAnchor),

      sceneView.topAnchor.constraint(equalTo: stage.topAnchor),
      sceneView.leadingAnchor.constraint(equalTo: stage.leadingAnchor),
      sceneView.trailingAnchor.constraint(equalTo: stage.trailingAnchor),
      sceneView.bottomAnchor.constraint(equalTo: stage.bottomAnchor),

      spinner.centerXAnchor.constraint(equalTo: stage.centerXAnchor),
      spinner.centerYAnchor.constraint(equalTo: stage.centerYAnchor),

      // 44pt minimum touch target.
      arButton.heightAnchor.constraint(greaterThanOrEqualToConstant: 44),
      arButton.centerXAnchor.constraint(equalTo: stage.centerXAnchor),
      arButton.bottomAnchor.constraint(equalTo: stage.bottomAnchor, constant: -16),
    ])
  }

  private func applyButtonStyle() {
    var configuration = UIButton.Configuration.filled()
    configuration.baseBackgroundColor = accentColor
    configuration.baseForegroundColor = .white
    configuration.cornerStyle = .capsule
    configuration.contentInsets = NSDirectionalEdgeInsets(top: 0, leading: 20, bottom: 0, trailing: 20)
    configuration.attributedTitle = AttributedString(
      arLabel, attributes: AttributeContainer([.font: UIFont.systemFont(ofSize: 15, weight: .semibold)]))
    arButton.configuration = configuration
    arButton.accessibilityLabel = arLabel
  }

  public override func didMoveToWindow() {
    super.didMoveToWindow()
    sceneView.isPlaying = window != nil && modelNode != nil && (item?.model?.viewSettings.autoRotate ?? false)
  }

  // MARK: Display

  /// Shows a dish. Pass `nil` to clear the view.
  public func display(_ item: EmbedItem?) {
    self.item = item
    presentation = resolvePresentation(platform: .iOS, model: item?.model)

    // The poster stays mounted beneath the canvas until the mesh reports in,
    // so there is never a blank frame between the two.
    let posterURL: String?
    if case .poster(let url) = presentation.inline {
      posterURL = url
    } else {
      posterURL = item?.model?.posterUrl ?? item?.imageUrl
    }
    loadPoster(posterURL)

    if case .usdz(let url) = presentation.inline, let settings = item?.model?.viewSettings {
      loadMesh(url, settings: settings)
    } else {
      clearMesh()
    }

    let offersAR = presentation.ar.supported && ArmenusAR.isSupported
    arButton.isHidden = !offersAR
    noteLabel.text = presentation.ar.reason
    noteLabel.isHidden = presentation.ar.reason == nil

    if offersAR, let usdz = item?.model?.usdzUrl {
      // Warm the cache as soon as AR is offerable rather than on tap. Quick
      // Look needs a local file either way; only the timing is ours.
      Task { _ = try? await ArmenusModelCache.shared.file(for: usdz) }
    }
  }

  private func loadPoster(_ urlString: String?) {
    posterTask?.cancel()
    posterView.image = nil
    guard let urlString, let url = URL(string: urlString) else { return }
    posterTask = Task { [weak self] in
      guard let (data, _) = try? await URLSession.shared.data(from: url), !Task.isCancelled else { return }
      let image = UIImage(data: data)
      await MainActor.run { self?.posterView.image = image }
    }
  }

  // MARK: Mesh

  private func loadMesh(_ source: String, settings: ModelViewSettings) {
    guard source != currentSource else { return }
    currentSource = source
    clearScene()
    spinner.startAnimating()

    ArmenusModelCache.shared.file(for: source) { [weak self] local, error in
      DispatchQueue.main.async {
        guard let self, self.currentSource == source else { return }
        guard let local else {
          self.spinner.stopAnimating()
          self.onModelError?(error?.localizedDescription ?? "Download failed")
          return
        }
        self.loadScene(from: local, settings: settings)
      }
    }
  }

  private func clearMesh() {
    currentSource = nil
    clearScene()
    spinner.stopAnimating()
  }

  private func clearScene() {
    modelNode?.removeAllActions()
    modelNode = nil
    sceneView.scene = nil
    sceneView.isPlaying = false
    sceneView.isHidden = true
  }

  private func loadScene(from url: URL, settings: ModelViewSettings) {
    // Parsed off the main thread. A textured dish is a few MB of USDZ and
    // decoding it inline drops frames in whatever list the card sits in.
    DispatchQueue.global(qos: .userInitiated).async { [weak self] in
      do {
        let scene = try SCNScene(url: url, options: [.checkConsistency: false])
        DispatchQueue.main.async {
          guard let self, self.currentSource != nil else { return }
          self.install(scene: scene, settings: settings)
        }
      } catch {
        DispatchQueue.main.async {
          self?.spinner.stopAnimating()
          self?.onModelError?(error.localizedDescription)
        }
      }
    }
  }

  private func install(scene: SCNScene, settings: ModelViewSettings) {
    sceneView.scene = scene

    let node = SCNNode()
    for child in scene.rootNode.childNodes {
      node.addChildNode(child)
    }
    scene.rootNode.addChildNode(node)
    modelNode = node

    // Re-centre on the bounding box. Exporters disagree about origin as wildly
    // as they do about scale; without this the camera can frame empty space.
    let (minBound, maxBound) = node.boundingBox
    let centre = SCNVector3(
      (minBound.x + maxBound.x) / 2,
      (minBound.y + maxBound.y) / 2,
      (minBound.z + maxBound.z) / 2)
    node.pivot = SCNMatrix4MakeTranslation(centre.x, centre.y, centre.z)

    applyCamera(CameraOrbit(parsing: settings.cameraOrbit))
    applyLight(exposure: settings.exposure, shadowIntensity: settings.shadowIntensity)
    applyAutoRotate(settings.autoRotate)

    sceneView.isHidden = false
    spinner.stopAnimating()
    onModelLoad?()
  }

  private func applyCamera(_ orbit: CameraOrbit) {
    guard let node = modelNode else { return }

    let (minBound, maxBound) = node.boundingBox
    let extent = SCNVector3(
      maxBound.x - minBound.x, maxBound.y - minBound.y, maxBound.z - minBound.z)
    let radius = max(extent.x, max(extent.y, extent.z)) / 2
    let distance = Float(orbit.radius) * radius * 2.6

    // Spherical to Cartesian, matching model-viewer's convention so that one
    // `cameraOrbit` string frames the dish identically on web and native.
    let theta = Float(orbit.theta) * .pi / 180
    let phi = Float(orbit.phi) * .pi / 180

    let camera = sceneView.pointOfView ?? {
      let cameraNode = SCNNode()
      cameraNode.camera = SCNCamera()
      sceneView.scene?.rootNode.addChildNode(cameraNode)
      sceneView.pointOfView = cameraNode
      return cameraNode
    }()

    camera.position = SCNVector3(
      distance * sin(phi) * sin(theta),
      distance * cos(phi),
      distance * sin(phi) * cos(theta))
    camera.look(at: SCNVector3Zero)
    camera.camera?.zNear = Double(max(0.01, distance * 0.01))
    camera.camera?.zFar = Double(distance * 10)
  }

  private func applyLight(exposure: Double, shadowIntensity: Double) {
    guard let scene = sceneView.scene else { return }

    let node = SCNNode()
    node.name = "armenus-key"
    let light = SCNLight()
    light.type = .directional
    light.intensity = CGFloat(exposure * 1000)
    light.castsShadow = shadowIntensity > 0
    light.shadowMode = .deferred
    light.shadowRadius = 8
    light.shadowColor = UIColor.black.withAlphaComponent(CGFloat(0.4 * shadowIntensity))
    node.light = light
    node.eulerAngles = SCNVector3(-Float.pi / 3, Float.pi / 5, 0)
    scene.rootNode.addChildNode(node)
  }

  private func applyAutoRotate(_ enabled: Bool) {
    guard let node = modelNode else { return }
    node.removeAction(forKey: "armenus-spin")

    guard enabled else {
      sceneView.isPlaying = false
      return
    }

    // One revolution every 18 seconds, matching the web and Android viewers.
    let spin = SCNAction.repeatForever(SCNAction.rotateBy(x: 0, y: .pi * 2, z: 0, duration: 18))
    node.runAction(spin, forKey: "armenus-spin")
    sceneView.isPlaying = window != nil
  }

  // MARK: AR

  @objc private func enterAR() {
    guard let item, let usdz = item.model?.usdzUrl, !presenting else { return }
    presenting = true
    onEnterAR?(item)

    Task { @MainActor [weak self] in
      do {
        // Life-size and fixed: the model is already scaled to the dish's real
        // dimensions, which is the whole question being asked.
        try await ArmenusAR.present(usdzURL: usdz, allowScaling: false)
      } catch {
        self?.onModelError?(error.localizedDescription)
      }
      self?.presenting = false
    }
  }
}
#endif

File: android/src/main/kotlin/app/armenus/sdk/ArmenusModelView.kt
package app.armenus.sdk

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Choreographer
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import io.github.sceneview.SceneView
import io.github.sceneview.math.Position
import io.github.sceneview.node.ModelNode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.URL
import java.nio.ByteBuffer
import kotlin.math.cos
import kotlin.math.sin

/**
 * A dish in 3D, with AR where the handset supports it.
 *
 * Filament, through SceneView, for the inline preview. GLB only: Filament has
 * no USDZ importer, the mirror of SceneKit's missing glTF one, and the reason
 * every model in this system is stored in both formats. No ARCore session is
 * created; AR is a handoff to Scene Viewer.
 *
 * The view is a square stage plus a one-line note beneath it, so give it a
 * width and let it size its height. In Jetpack Compose, wrap it in
 * `AndroidView` and call [display] from the `update` block.
 */
class ArmenusModelView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
) : LinearLayout(context, attrs) {

  /* -- configuration --------------------------------------------------------- */

  var arLabel: String = "View on your table"
    set(value) {
      field = value
      arButton.text = value
    }

  /**
   * Whether a drag rotates the model. Pass false inside a RecyclerView: on a
   * phone the two gesture systems fight, and a card that swallows vertical
   * drags makes the whole feed feel stuck.
   */
  var interactionEnabled: Boolean = true

  /** Background of the AR button. */
  var accentColor: Int = Color.parseColor("#234D3C")
    set(value) {
      field = value
      applyButtonStyle()
    }

  var onModelLoad: (() -> Unit)? = null
  var onModelError: ((String) -> Unit)? = null
  var onEnterAr: ((EmbedItem) -> Unit)? = null

  var item: EmbedItem? = null
    private set

  var presentation: Presentation = resolvePresentation(Platform.ANDROID, null)
    private set

  /* -- views ----------------------------------------------------------------- */

  private val stage = Stage(context)
  private val posterView = ImageView(context)
  private val progress = ProgressBar(context)
  private val arButton = Button(context)
  private val note = TextView(context)
  private var sceneView: SceneView? = null

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private var modelNode: ModelNode? = null
  private var currentSource: String? = null
  private var posterJob: Job? = null
  private var loadJob: Job? = null

  init {
    orientation = VERTICAL

    stage.background = GradientDrawable().apply {
      setColor(Color.parseColor("#F4F2F3"))
      cornerRadius = dp(14f)
    }
    stage.clipToOutline = true

    posterView.scaleType = ImageView.ScaleType.FIT_CENTER
    stage.addView(posterView, FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT))

    progress.isIndeterminate = true
    progress.visibility = View.GONE
    stage.addView(progress, FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT, Gravity.CENTER))

    arButton.text = arLabel
    arButton.isAllCaps = false
    arButton.setTextColor(Color.WHITE)
    arButton.setTextSize(TypedValue.COMPLEX_UNIT_SP, 15f)
    arButton.minHeight = dp(44f).toInt()
    arButton.minimumHeight = dp(44f).toInt()
    arButton.setPadding(dp(20f).toInt(), 0, dp(20f).toInt(), 0)
    arButton.visibility = View.GONE
    arButton.contentDescription = arLabel
    arButton.setOnClickListener { enterAr() }
    applyButtonStyle()
    stage.addView(
      arButton,
      FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT, Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL).apply {
        bottomMargin = dp(16f).toInt()
      },
    )

    addView(stage, LayoutParams(MATCH_PARENT, WRAP_CONTENT))

    note.setTextSize(TypedValue.COMPLEX_UNIT_SP, 13f)
    note.setTextColor(Color.parseColor("#6B5555"))
    note.gravity = Gravity.CENTER_HORIZONTAL
    note.visibility = View.GONE
    addView(note, LayoutParams(MATCH_PARENT, WRAP_CONTENT).apply { topMargin = dp(10f).toInt() })
  }

  /* -- display --------------------------------------------------------------- */

  /** Shows a dish. Pass null to clear the view. */
  fun display(item: EmbedItem?) {
    this.item = item
    presentation = resolvePresentation(
      platform = Platform.ANDROID,
      model = item?.model,
      arCoreAvailable = ArmenusAr.isAvailable(context),
    )

    // The poster stays mounted beneath the canvas until the mesh reports in,
    // so there is never a blank frame between the two.
    val inline = presentation.inline
    val posterUrl = (inline as? InlineSource.Poster)?.url ?: item?.model?.posterUrl ?: item?.imageUrl
    loadPoster(posterUrl)

    if (inline is InlineSource.Glb && item?.model != null) {
      loadMesh(inline.url, item.model.viewSettings)
    } else {
      clearMesh()
    }

    arButton.visibility = if (presentation.ar.supported) View.VISIBLE else View.GONE
    note.text = presentation.ar.reason
    note.visibility = if (presentation.ar.reason == null) View.GONE else View.VISIBLE
  }

  private fun loadPoster(url: String?) {
    posterJob?.cancel()
    posterView.setImageDrawable(null)
    if (url.isNullOrEmpty()) return
    posterJob = scope.launch {
      val bitmap = runCatching {
        withContext(Dispatchers.IO) {
          URL(url).openStream().use { BitmapFactory.decodeStream(it) }
        }
      }.getOrNull()
      if (bitmap != null) posterView.setImageBitmap(bitmap)
    }
  }

  /* -- mesh ------------------------------------------------------------------ */

  private fun loadMesh(source: String, settings: ModelViewSettings) {
    if (source == currentSource) return
    currentSource = source
    clearScene()
    progress.visibility = View.VISIBLE

    loadJob?.cancel()
    loadJob = scope.launch {
      try {
        // Fetched and read off the main thread. A textured dish is a few MB,
        // and doing it inline drops frames in whatever list the card is
        // scrolling in, the most visible performance mistake on this path.
        val bytes = withContext(Dispatchers.IO) {
          ArmenusModelCache.fileBlocking(context, source).readBytes()
        }
        if (currentSource != source) return@launch

        val view = sceneView ?: SceneView(context).also {
          sceneView = it
          stage.addView(it, 1, FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT))
        }

        val instance = view.modelLoader.createModelInstance(ByteBuffer.wrap(bytes))
        val node = ModelNode(
          modelInstance = instance,
          autoAnimate = false,
          // Normalised to a unit bounding box so `cameraOrbit` means the same
          // thing for a coffee cup and a sharing platter. Real-world scale is
          // applied by the AR viewer, not by the preview.
          scaleToUnits = 1.0f,
          centerOrigin = Position(0f, 0f, 0f),
        )
        view.addChildNode(node)
        modelNode = node

        applyCamera(view, CameraOrbit.parse(settings.cameraOrbit))
        if (settings.autoRotate) startSpin() else stopSpin()

        progress.visibility = View.GONE
        onModelLoad?.invoke()
      } catch (error: Throwable) {
        if (error is kotlinx.coroutines.CancellationException) throw error
        progress.visibility = View.GONE
        onModelError?.invoke(error.message ?: "Model failed to load")
      }
    }
  }

  private fun clearMesh() {
    currentSource = null
    loadJob?.cancel()
    clearScene()
    progress.visibility = View.GONE
  }

  private fun clearScene() {
    stopSpin()
    val view = sceneView
    modelNode?.let { node ->
      view?.removeChildNode(node)
      node.destroy()
    }
    modelNode = null
  }

  private fun applyCamera(view: SceneView, orbit: CameraOrbit) {
    // Spherical to Cartesian, matching model-viewer's convention so that one
    // `cameraOrbit` string frames a dish identically on web, iOS and Android.
    val theta = Math.toRadians(orbit.theta)
    val phi = Math.toRadians(orbit.phi)
    val distance = orbit.radius * 2.6

    view.cameraNode.position = Position(
      x = (distance * sin(phi) * sin(theta)).toFloat(),
      y = (distance * cos(phi)).toFloat(),
      z = (distance * sin(phi) * cos(theta)).toFloat(),
    )
    view.cameraNode.lookAt(targetWorldPosition = Position(0f, 0f, 0f), smooth = false)
  }

  /* -- spin ------------------------------------------------------------------ */

  private var spinCallback: Choreographer.FrameCallback? = null

  private fun startSpin() {
    stopSpin()
    val node = modelNode ?: return
    var last = 0L

    // Driven by Choreographer rather than a fixed timer, so the rotation is
    // tied to actual frame delivery. A timer keeps firing while the view is
    // off screen or the device is throttling, which burns battery and makes
    // the spin jump when it comes back.
    val callback = object : Choreographer.FrameCallback {
      override fun doFrame(frameTimeNanos: Long) {
        if (last != 0L) {
          val delta = (frameTimeNanos - last) / 1_000_000_000f
          // One revolution every 18 seconds, matching the web and iOS viewers.
          node.rotation = node.rotation.copy(y = node.rotation.y + delta * (360f / 18f))
        }
        last = frameTimeNanos
        if (spinCallback === this) Choreographer.getInstance().postFrameCallback(this)
      }
    }
    spinCallback = callback
    Choreographer.getInstance().postFrameCallback(callback)
  }

  private fun stopSpin() {
    spinCallback?.let { Choreographer.getInstance().removeFrameCallback(it) }
    spinCallback = null
  }

  /* -- AR -------------------------------------------------------------------- */

  private fun enterAr() {
    val current = item ?: return
    val glb = current.model?.glbUrl ?: return
    onEnterAr?.invoke(current)
    // Life-size and fixed: the model is already scaled to the dish's real
    // dimensions, which is the whole question being asked.
    if (!ArmenusAr.launch(context, glb, title = current.name, allowScaling = false)) {
      onModelError?.invoke("Scene Viewer could not be launched")
    }
  }

  /* -- lifecycle ------------------------------------------------------------- */

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    // Filament holds native GPU resources that the GC will not reclaim, so an
    // undestroyed SceneView per scrolled-past card leaks until the app dies.
    // Call `display(item)` again after re-attaching, as a RecyclerView binder does.
    stopSpin()
    scope.coroutineContext.cancelChildren()
    clearScene()
    sceneView?.let { view ->
      stage.removeView(view)
      view.destroy()
    }
    sceneView = null
    currentSource = null
  }

  /* -- helpers --------------------------------------------------------------- */

  private fun applyButtonStyle() {
    arButton.background = GradientDrawable().apply {
      setColor(accentColor)
      cornerRadius = dp(999f)
    }
    arButton.stateListAnimator = null
  }

  private fun dp(value: Float): Float =
    TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, value, resources.displayMetrics)

  /**
   * The square canvas. When interaction is off it lets touches fall through
   * to the enclosing list instead of the renderer's own gesture handling,
   * while still delivering taps to the AR button.
   */
  private inner class Stage(context: Context) : FrameLayout(context) {
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
      val width = MeasureSpec.getSize(widthMeasureSpec)
      val square = MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY)
      super.onMeasure(square, square)
    }

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
      if (interactionEnabled) return false
      val onButton = arButton.visibility == View.VISIBLE &&
        event.x >= arButton.left && event.x <= arButton.right &&
        event.y >= arButton.top && event.y <= arButton.bottom
      return !onButton
    }

    override fun onTouchEvent(event: MotionEvent): Boolean = false
  }

  private companion object {
    const val MATCH_PARENT = LayoutParams.MATCH_PARENT
    const val WRAP_CONTENT = LayoutParams.WRAP_CONTENT
  }
}
